cpg_workflows/large_cohort/scripts/ancestry_add_background.py

```diff
@@ -23,15 +23,14 @@ def prepare_background_mt(data_path: str, qc_variants_ht: hl.Table, tmp_path) ->
     Returns:
         A densified MT, checkpointed
     """
-    background_mt_checkpoint_path = join(tmp_path, 'densified_background.mt')
+    background_mt_checkpoint_path = join(tmp_path, f'{dataset}_densified_background.mt')
```

Comment:
I'm actually kinda keen to cut this prep stage out into a separate Stage - it looks like we read these MT/VDSs in, do some stuff on them, and write them into temp... then next time the pipeline runs we do it all over again. Why not make a stage where we prepare each background MT as a densified MT, save it to long term storage, then we don't need to do this each time.

Are these the 1+TB tables?

Reply:
Yep, the background vds for hgdp-1kg is around 1.6Tib (`gs://cpg-hgdp-1kg-test/vds-main.vds`). We ran into issues with this a couple of months ago ([Slack here](https://centrepopgen.slack.com/archives/C030X7WGFCL/p1712720352383989)).

I added the saving to `tmp` not that long ago because it saved time during debugging. I guess that these runs on main are somewhat infrequent and so storing (what could be a fairly large file) may be reason to avoid more permanent storage. That's just me being naive to how much it costs to store things and how big the files for may be. Suffice to say, I think it would be good to have a more permanent storage for them, if anything to inspect for accuracy and quality control

Reply:
To be fair, I don't know how many different types of background datasets we want to add/will have access to. This may very well be the only one that suits our needs. Which would mean it's okay to/we probably should store permanent intermediates of this dataset

Reply:
Cool cool. I wasn't sure how many background tables exist or how large they are. If it's one fat table that we reprocess with each (infrequent) run that's something we can plug into the ol' calculator and make a decision about. I haven't seen this stage run on main data, so not sure how long this is expected to take, or how much it cost in dataproc

Reply:
So I looked into possibly splitting this out as it's separate stage. Couple of things:
- The dataproc removal process is already adding stages and I'm not sure how this sits within the PA team, so at this point if we can prevent adding another stage that would be preferential
- We currently have a `DenseSubset` stage which essentially does what we want it to but for the dataset in question (i.e. `bioheart`, `tob-wgs`, `tenk10k` and not the background datasets). If we wanted to do it for the background dataset then it would require a completely separate pipeline run to run just that stage which kind of breaks the autonomy.
- I think we can implement a check with within the `prepare_background()` function (with possibly a another config parameter 🙄) for the existence of the background datasets densified matrix table. This would mean that it would all happen within the current pipeline run

Reply:
```suggestion
    bg_path = config_retrieve(['storage', bg_dataset, 'default'])
    background_mt_checkpoint_path = join(bg_path, 'dense_background.mt')
    if can_reuse(background_mt_checkpoint_path):
        return hl.read_matrix_table(background_mt_checkpoint_path)
```

Reply:
> The dataproc removal process is already adding stages and I'm not sure how this sits within the PA team, so at this point if we can prevent adding another stage that would be preferential

From my POV you're thinking about stage splitting all wrong - the current process is super complicated, takes tons of inputs, writes tons of outputs, and not in a way that's easily disentangled. All the individual parts of the process run independently, but all one after another in a single Stage, so changing any part of the process means changing the whole Stage. 

The most important part here is that the script as written includes a number of `resume-from-checkpoint` statements, so a new run of the 'Relatedness' or 'Ancestry' Stage doesn't mean you're running the whole process, it means the method can resume from a previous failed run. That failure might have been from a previous run or a previous commit you don't want to continue from. I heavily advocate against any statements which allow `this` run to continue from `that` run under any circumstances.

Splitting this one trash stage into several stages means that each step of the process is discrete - only take the relevant inputs for `this` part of the process, and do the minimum amount of work to achieve that. Once it's implemented we have multiple simple steps instead of one monolithic step, but the pipeline runs automatically, so that doesn't cause any more effort to operate, and it's much easier to debug.

Reply:
> If we wanted to do it for the background dataset then it would require a completely separate pipeline run to run just that stage which kind of breaks the autonomy.

I don't agree here. The pipeline stages only run if their outputs don't exist. If you create a stage called `MakeDenseHGDP` (for example) which writes a densified version of the HGDP MT into permanent storage (maybe CPG common?), you can add that to the start of the pipeline and then use the outputs in other stages. Subsequent runs of the pipeline would check that it exists, and not run that stage, but would be able to re-use the output. That wouldn't have to be a separate pipeline/workflow in order to work.

I'm not convinced that it needs to be a Stage of its own, but if it were to be broken out like that I would create it as a Stage within the large_cohort pipeline, and only expect it to ever run once (unless the logic of how to generate the BG datasets changes)